src/client/debugger/jupyter/debuggingManager.ts

```diff
@@ -92,102 +93,119 @@ export class DebuggingManager implements IExtensionSingleActivationService, IDeb
 
-            this.commandManager.registerCommand(DSCommands.RunByLine, async (cell: NotebookCell | undefined) => {
-                sendTelemetryEvent(DebuggingTelemetry.clickedRunByLine);
-                const editor = this.vscNotebook.activeNotebookEditor;
-                if (!cell) {
-                    const range = editor?.selections[0];
-                    if (range) {
-                        cell = editor?.document.cellAt(range.start);
+            this.commandManager.registerCommand(
+                DSCommands.RunByLine,
+                throttle(async (cell: NotebookCell | undefined) => {
+                    sendTelemetryEvent(DebuggingTelemetry.clickedRunByLine);
+                    const editor = this.vscNotebook.activeNotebookEditor;
+                    if (!cell) {
+                        const range = editor?.selections[0];
+                        if (range) {
+                            cell = editor?.document.cellAt(range.start);
+                        }
                     }
-                }
 
-                if (!cell) {
-                    return;
-                }
+                    if (!cell) {
+                        return;
+                    }
 
-                if (editor) {
-                    if (await this.checkForIpykernel6(editor.document, DataScience.startingRunByLine())) {
-                        this.updateToolbar(true);
-                        this.updateCellToolbar(true);
-                        await this.startDebuggingCell(editor.document, KernelDebugMode.RunByLine, cell);
+                    if (editor) {
+                        if (await this.checkForIpykernel6(editor.document, DataScience.startingRunByLine())) {
+                            this.updateToolbar(true);
+                            this.updateCellToolbar(true);
+                            await this.startDebuggingCell(editor.document, KernelDebugMode.RunByLine, cell);
+                        } else {
+                            void this.installIpykernel6();
+                        }
                     } else {
-                        void this.installIpykernel6();
+                        void this.appShell.showErrorMessage(DataScience.noNotebookToDebug());
                     }
-                } else {
-                    void this.appShell.showErrorMessage(DataScience.noNotebookToDebug());
-                }
-            }),
-
-            this.commandManager.registerCommand(DSCommands.RunByLineNext, (cell: NotebookCell | undefined) => {
-                if (!cell) {
-                    const editor = this.vscNotebook.activeNotebookEditor;
-                    const range = editor?.selections[0];
-                    if (range) {
-                        cell = editor?.document.cellAt(range.start);
+                }, 1000)
+            ),
+
+            this.commandManager.registerCommand(
+                DSCommands.RunByLineNext,
+                throttle((cell: NotebookCell | undefined) => {
```

Comment:
Why throttle this? When I am doing normal javascript debugging, I can step as fast as I want. I can even hold down F10 to step really fast. The same should work here.

Reply:
Because that seems to be the issue, clicking fast enough causes debugpy to freeze.

The changes to the controller are what gets it out of the situation (finishes that frozen session and un freezes debugpy).
The throttling is meant to prevent the situation.

Its removable, but I'd rather have at least some.

Reply:
It's hard to believe that debugpy just can't step fast. I can debug a normal python script and step over 50 lines in a couple seconds

Reply:
I agree, I think we need to identify the root cause, this feels like a bandaid

Reply:
I think they problem is not because it happens fast, it's because we start multiple debug sessions instead of checkingbif we have already started one